src/phoenix/experimental/evals/models/rate_limiters.py

```diff
@@ -0,0 +1,230 @@
+import asyncio
+import time
+from collections import defaultdict
+from functools import wraps
+from typing import Any, Awaitable, Callable, Dict, Optional, TypeVar, Union, cast
+
+from typing_extensions import ParamSpec
+
+Numeric = Union[int, float]
+ParameterSpec = ParamSpec("ParameterSpec")
+GenericType = TypeVar("GenericType")
+AsyncCallable = TypeVar("AsyncCallable", bound=Callable[..., Awaitable[Any]])
+
+
+class UnavailableTokensError(Exception):
+    pass
+
+
+class TokenRateLimiter:
+    """
+    A in-memory rate-limiter implemented using the token bucket algorithm.
+
+    A description of the technique can be found here: (https://en.wikipedia.org/wiki/Token_bucket).
+    A collection of arbitrary tokens replenish at `rate` tokens per second up to a maximum. These
+    tokens can be spent to process requests.
+
+    Args:
+    per_minute_rate (float): The rate at which tokens replenish.
+
+    starting_tokens (float): The number of tokens the rate limiter starts with.
+
+    max_tokens (float): The maximum number of tokens this rate limiter can hold. Tokens will not
+    replenish past this value.
+
+    rate_multiplier (float, Optional): A multiplier used to adjust the rate limit applied by this
+    object. For example, a rate_multiplier of 0.8 can be used to ensure that all actions
+    controlled by this rate limiter will never exceed 0.8 times the originally specified rate.
+    """
+
+    def __init__(
+        self,
+        per_second_rate: Numeric,
+        starting_tokens: Numeric,
+        max_tokens: Numeric,
+        rate_multiplier: Numeric = 1,
```

Comment:
In a sense this parameter is redundant because the user can just do the multiplication themselves, but now that there are two ways to do the same thing, it can be a source of confusion.